tools/eslint-plugin/dist/rules/public-extension-exports.js

```diff
@@ -0,0 +1,222 @@
+        declaration.declarationList.declarations[0].symbol.escapedName :
+        declaration.symbol.escapedName;
+      const kind = getSyntaxKindFriendlyName(declaration.kind);
+      const csvString = createCsvString(name, kind);
+      fs.writeFileSync(apiFilePath, csvString, { flag: "a" });
+    }
+
+    function getFileName(parent) {
+      let currentParent = parent;
+      while (currentParent) {
+        if (currentParent.fileName !== undefined)
+          return currentParent.fileName;
+        currentParent = currentParent.parent;
+      }
+      return undefined;
+    }
+
+    function isLocalFile(declaration) {
+      if (declaration) {
+        const fileName = getFileName(declaration.parent);
+        if (fileName && typeof fileName === "string" && !fileName.includes("node_modules"))
+          return true;
+      }
+      return false;
+    }
+
+    function getParentSymbolName(declaration) {
+      if (declaration.parent && declaration.parent.symbol && !declaration.parent.symbol.escapedName.startsWith('"'))
+        return declaration.parent.symbol.escapedName;
+      return undefined;
+    }
+
+    function checkJsDoc(declaration, node) {
+      // Only check local elements, not consumed ones
+      if (!declaration || !declaration.jsDoc || !isLocalFile(declaration))
+        return undefined;
+
+      for (const jsDoc of declaration.jsDoc)
+        if (jsDoc.tags) {
+          let jsDocExtensionTag = jsDoc.tags.find(tag => tag.tagName.escapedText === extensionsTag);
+          let jsDocPreviewTag = jsDoc.tags.find(tag => tag.tagName.escapedText === previewTag);
+          // Has extension API tag
+          if (jsDocExtensionTag) {
+            addToApiList(declaration, jsDocPreviewTag);
+            // Does not have any of the required release tags
+            if (!jsDoc.tags.some(tag => releaseTags.includes(tag.tagName.escapedText))) {
+              let name;
+              if (declaration.kind === ts.SyntaxKind.Constructor)
+                name = declaration.parent.symbol.escapedName;
+              else {
+                name = declaration.symbol.escapedName;
+                const parentSymbol = getParentSymbolName(declaration);
+                if (parentSymbol)
+                  name = `${parentSymbol}.${name}`;
+              }
+
+              context.report({
+                node,
+                messageId: "forbidden",
+                data: {
+                  kind: getSyntaxKindFriendlyName(declaration.kind),
+                  name,
+                  releaseTags: releaseTags,
+                }
+              });
+            }
+          }
+        }
+    }
+
+    function checkWithParent(declaration, node) {
+      if (!declaration)
+        return;
+      checkJsDoc(declaration, node);
+      if (declaration.parent && [
+        ts.SyntaxKind.ClassDeclaration,
+        ts.SyntaxKind.EnumDeclaration,
+        ts.SyntaxKind.InterfaceDeclaration,
+        ts.SyntaxKind.ModuleDeclaration,
```

Comment:
is this an exhaustive list of node types that can have release tags? I'm not sure if it's even worth filtering by `SyntaxKind`, the Typescript AST might have edge cases

Reply:
The filtering only applies to the node's parent, but if we shouldn't be checking parents, then no need for this code.

Reply:
yeah I agree, but by design `@public` tags are inheritable, at least by api-extractor's rules. So I think it should check parents (and further ancestors), but it shouldn't filter them by type.